src/main/java/dev/blaauwendraad/masker/json/MaskingState.java

```diff
@@ -250,19 +293,86 @@ private void checkCurrentValueBounds(int index) {
         }
     }
 
+    /**
+     * Flushes the current buffer into provided OutputStream, then moves the current token to the beginning of the buffer
+     * and fills up the buffer from provided InputStream.
+     * In case the current token is too long (the start index is not in the last quarter of the buffer),
+     * extends the current buffer by a quarter.
+     *
+     * @return true if more data is available in the input stream, false if the end of the stream is reached.
+     */
+    boolean readNextBuffer() {
+        if (inputStream == null || outputStream == null) {
+            return false;
+        }
+
+        // flush the remaining of the current buffer up to the current value start index into output stream
+        try {
+            int end = currentValueStartIndex == -1 ? messageLength : currentValueStartIndex;
+            outputStream.write(message, lastReplacementEndIndex, end - lastReplacementEndIndex);
+            outputStream.flush();
```

Comment:
Why do we need flush here?

Reply:
After I remove the flush in "replaceValue" method, this will be the only flush that we do to the output stream. I think it makes sense that we flush "physical" output stream buffer while we read the next "logical" masker buffer.

Reply:
Agreed

Reply:
Done